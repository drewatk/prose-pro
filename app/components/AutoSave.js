import React from "react";
import { connect } from "react-redux";
import { convertToRaw } from "draft-js";

const AUTOSAVE_DELAY_MS = 5 * 1000;

export class AutoSave extends React.Component {
  componentDidMount() {
    this.interval = setInterval(() => {
      this.save();
    }, AUTOSAVE_DELAY_MS);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  save() {
    const { gitAbstractions, currentFile, editorState } = this.props;

    if (gitAbstractions && currentFile && editorState) {
      gitAbstractions.saveFile(
        currentFile,
        convertToRaw(editorState.getCurrentContent())
      );

      console.log("Editor saved!");
    } else {
      console.warn("Editor failed to save");
    }
  }

  render() {
    return this.props.children;
  }
}

const mapStateToProps = ({
  editor: { editorState },
  currentFile,
  gitAbstractions
}) => {
  return {
    currentFile,
    gitAbstractions,
    editorState
  };
};

const mapDispatchToProps = dispatch => ({ dispatch });

const WithAutoSave = connect(
  mapStateToProps,
  mapDispatchToProps
)(AutoSave);

export default WithAutoSave;
