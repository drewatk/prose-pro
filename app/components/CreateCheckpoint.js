import React, { Component } from 'react';

import { Button, Input } from './Basic';

const styles = {
  form: {
    display: 'flex',
    flexDirection: 'column',
    height: '200px',
    width: '350px',
    backgroundColor: '#fff',
    borderWidth: '1px',
    borderColor: '#000',
    borderStyle: 'solid',
    alignItems: 'center',
    justifyContent: 'center',
    color: '#000'
  },
  button: {
    marginTop: '20px',
    width: '60px'
  }
};

/* I don't really want this to be a separate component in the file structure.
   It should be written as a separate component direcly inside the file that uses it */
class CreateCheckpoint extends Component {
  constructor(props) {
    super(props);
    this.state = { commitMessage: '' };
  }

  handleTextChange(e) {
    this.setState({ commitMessage: e.target.value });
  }

  handleSubmit(e) {
    e.preventDefault();
    console.log(this.state.commitMessage);
  }

  render() {
    return (
      <form style={styles.form} onSubmit={this.handleSubmit.bind(this)}>
        {/* Set both button and input to disabled if no file changes have been saved. */}
        <h4>Create Checkpoint</h4>
        <label>Checkpoint Message (required)</label>
        <Input
          value={this.state.commitMessage}
          onChange={this.handleTextChange.bind(this)}
        />
        <Button type="submit" text="Create" style={styles.button} />
      </form>
    );
  }
}

export default CreateCheckpoint;
