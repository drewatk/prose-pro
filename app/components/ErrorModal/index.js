import React from "react";
import { connect } from "react-redux";
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from "reactstrap";
import { hideError } from "app/actions/error";

export const ErrorModal = props => {
  const { hasError, errorMessage, children, dispatch } = props;

  return (
    <React.Fragment>
      {children}
      <Modal isOpen={hasError}>
        <ModalHeader>Error</ModalHeader>
        <ModalBody>{errorMessage}</ModalBody>
        <ModalFooter>
          <Button color="danger" onClick={() => dispatch(hideError())}>
            Close
          </Button>
        </ModalFooter>
      </Modal>
    </React.Fragment>
  );
};

const mapStateToProps = state => {
  return {
    hasError: state.error.hasError,
    errorMessage: state.error.errorMessage
  };
};
const mapDispatchToProps = dispatch => ({ dispatch });

const WithErrorModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(ErrorModal);

WithErrorModal.displayName = "ErrorModal";
export default WithErrorModal;
