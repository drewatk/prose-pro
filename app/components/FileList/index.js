import React from "react";
import { connect } from "react-redux";

import CreateFileForm from "../Forms/CreateFileForm";
import FileNameList from "./FileNameList";

import selectFile from "app/actions/file_selection";
import updateFiles from "app/actions/files";

const FileList = ({ files, gitAbstractions, dispatch }) => {
  return (
    <div>
      <div>
        <CreateFileForm
          onSubmit={({ fileName }) => {
            gitAbstractions
              .createFile(fileName)
              .then(() => dispatch(updateFiles(gitAbstractions.getFiles())))
              .then(() => dispatch(selectFile(gitAbstractions, fileName)))
              .catch(err =>
                console.error("Error in CreateFileForm onSubmit: ", err)
              );
          }}
        />
      </div>
      <div style={{ backgroundColor: "lightblue", height: "100%" }}>
        FileList View
      </div>
      <div>
        <FileNameList
          files={files}
          onFileItemClick={file => {
            dispatch([
              selectFile(gitAbstractions, file)
              /* load file data */
              /* load file checkpoints */
            ]);
          }}
        />
      </div>
    </div>
  );
};

const mapStateToProps = ({ files, gitAbstractions }) => ({
  files,
  gitAbstractions
});

const mapDispatchToProps = dispatch => ({ dispatch });

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FileList);
