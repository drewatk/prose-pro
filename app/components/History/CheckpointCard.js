import React from "react";
import { connect } from "react-redux";
import * as R from "ramda";
import {
  Card,
  CardBody,
  CardSubtitle,
  CardText,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";
import moment from "moment";

import styles from "./CheckpointCard.scss";

import { EditorState, convertFromRaw } from "draft-js";
import { stateToHTML } from "draft-js-export-html";

import diff from "app/utils/diff";
import {
  UPDATE_EDITOR_STATE,
  SET_VIEW_STATE,
  SET_EDIT_STATE,
  SET_DIFF_STATE
} from "app/actions/editor";
import updateHistory from "app/actions/history";

import { FileObject } from "app/git-abs/metadata/file-object";

import { showError } from "app/actions/error";

export class CheckpointCard extends React.Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.state = {
      dropdownOpen: false
    };
  }

  toggle() {
    this.setState(prevState => ({
      dropdownOpen: !prevState.dropdownOpen
    }));
  }

  render() {
    const {
      version,
      timestamp,
      commit,
      gitAbstractions,
      currentFile,
      checkpointHistory,
      viewedCheckpoint,
      dispatch
    } = this.props;

    return (
      <Card
        className={`${styles.card} ${
          viewedCheckpoint === commit ? styles.cardViewed : ""
        }`}
        data-test-id="history-list-card"
      >
        <CardBody>
          <CardSubtitle>Checkpoint:</CardSubtitle>
          <CardText>{version}</CardText>
          <CardSubtitle>Date:</CardSubtitle>
          <CardText>{moment(timestamp).fromNow()}</CardText>
          <Dropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
            <DropdownToggle tag="span">
              <i
                className="fas fa-ellipsis-h"
                style={{ float: "right" }}
                data-test-id="history-list-card-toggle"
              />
            </DropdownToggle>
            <DropdownMenu>
              <DropdownItem
                onClick={() => {
                  gitAbstractions
                    .switchToCurrentVersion(currentFile)
                    .then(() =>
                      gitAbstractions.switchVersion(currentFile, commit)
                    )
                    .then(fileData =>
                      dispatch({
                        type: UPDATE_EDITOR_STATE,
                        payload: EditorState.createWithContent(
                          convertFromRaw(fileData)
                        )
                      })
                    )
                    .then(() =>
                      gitAbstractions.switchToCurrentVersion(currentFile)
                    )
                    .then(() =>
                      dispatch({ type: SET_VIEW_STATE, payload: commit })
                    )
                    .catch(e => {
                      console.error("Error in Checkpoint View: ", e);
                      dispatch(showError(e.message));
                    });
                }}
                data-test-id="history-list-card-view-button"
              >
                View
              </DropdownItem>
              <DropdownItem
                onClick={() => {
                  gitAbstractions
                    .reset(currentFile, commit)
                    .then(fileData =>
                      dispatch({
                        type: UPDATE_EDITOR_STATE,
                        payload: EditorState.createWithContent(
                          convertFromRaw(fileData)
                        )
                      })
                    )
                    .then(() => dispatch({ type: SET_EDIT_STATE }))
                    .then(() =>
                      gitAbstractions.updateVersions(
                        currentFile,
                        new FileObject({
                          versions: R.dropLastWhile(
                            cp => cp.commit !== commit,
                            checkpointHistory
                          )
                        })
                      )
                    )
                    .then(() => gitAbstractions.getVersions(currentFile))
                    .then(({ versions }) => dispatch(updateHistory(versions)))
                    .catch(err => {
                      console.error("Reset Failed: ", err);
                      dispatch(showError(err.message));
                    });
                }}
                data-test-id="history-list-card-revert-button"
              >
                Revert
              </DropdownItem>
              <DropdownItem
                onClick={() => {
                  let selectedVersion = null,
                    previousVersion = null;
                  gitAbstractions
                    .switchToCurrentVersion(currentFile)
                    .then(() =>
                      gitAbstractions.switchVersion(currentFile, commit)
                    )
                    .then(selectedVersionData => {
                      selectedVersion = stateToHTML(
                        convertFromRaw(selectedVersionData)
                      );
                      return gitAbstractions.getVersions(currentFile);
                    })
                    .then(({ versions }) => {
                      const previousCommit = R.last(
                        R.takeWhile(v => v.commit !== commit, versions)
                      );
                      if (previousCommit) {
                        return gitAbstractions
                          .switchToCurrentVersion(currentFile)
                          .then(() =>
                            gitAbstractions.switchVersion(
                              currentFile,
                              previousCommit.commit
                            )
                          );
                      }
                      // return an empty file
                      return null;
                    })
                    .then(previousVersionData => {
                      if (previousVersionData === null) {
                        previousVersion = "";
                      } else {
                        previousVersion = stateToHTML(
                          convertFromRaw(previousVersionData)
                        );
                      }

                      return gitAbstractions.switchToCurrentVersion(
                        currentFile
                      );
                    })
                    .then(() =>
                      dispatch({
                        type: SET_DIFF_STATE,
                        payload: diff(previousVersion, selectedVersion).join("")
                      })
                    )
                    .catch(e => console.error("Error in Checkpoint Diff: ", e));
                }}
                data-test-id="history-list-card-diff-button"
              >
                Diff
              </DropdownItem>
            </DropdownMenu>
          </Dropdown>
        </CardBody>
      </Card>
    );
  }
}

const mapStateToProps = ({
  gitAbstractions,
  currentFile,
  checkpointHistory,
  editor: { viewedCheckpoint }
}) => ({
  gitAbstractions,
  currentFile,
  checkpointHistory,
  viewedCheckpoint
});

export default connect(mapStateToProps)(CheckpointCard);
