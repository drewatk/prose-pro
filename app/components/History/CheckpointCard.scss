.card {
  border: none !important;
  background: var(--dark-white-color) !important;
  color: var(--accent-color) !important;
  font-size: 1.2rem !important;
  margin-bottom: 5px;
}

.card-viewed {
  background: var(--accent-color) !important;
  color: var(--white-color) !important;
}

:global(.card-body) {
  padding: 10px !important;
}

:global(.card-subtitle) {
  margin-top: 0 !important;
  font-weight: bold !important;
}

:global(.card-text) {
  margin-bottom: 5px !important;
  font-size: 0.9rem !important;
}

:global(.dropdown-menu) {
  background-color: var(--accent-color) !important;
  border-color: var(--dark-white-color) !important;
}

:global(.dropdown-item) {
  color: var(--white-color) !important;

  &:hover {
    background-color: var(--dark-white-color) !important;
    color: var(--dark-color) !important;
  }
}
