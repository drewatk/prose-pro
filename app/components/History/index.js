import React from "react";
import { connect } from "react-redux";

import CheckpointView from "./CheckpointView";

const History = ({ history }) => (
  <div style={{ height: "100%" }}>
    {history.map((data, index) => (
      <CheckpointView key={index} {...data} />
    ))}
  </div>
);

const mapStateToProps = ({ checkpointHistory }) => ({
  history: checkpointHistory
});

export default connect(mapStateToProps)(History);
