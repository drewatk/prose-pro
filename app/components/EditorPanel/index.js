import React from "react";
import { connect } from "react-redux";
import ProseEditor from "app/components/ProseEditor";
import Viewer from "./Viewer";
import AutoSave from "app/components/AutoSave";

export const EditorPanel = props => {
  const { isEditable } = props;

  return (
    <React.Fragment>
      {isEditable && (
        <AutoSave>
          <ProseEditor />
        </AutoSave>
      )}
      {!isEditable && <Viewer />}
    </React.Fragment>
  );
};

EditorPanel.displayName = "EditorPanel";

const mapStateToProps = ({ editor: { isEditable } }) => ({ isEditable });
const WithEditorPanel = connect(mapStateToProps)(EditorPanel);

export default WithEditorPanel;
