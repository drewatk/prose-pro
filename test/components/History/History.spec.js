import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { History } from "app/components/History";

Enzyme.configure({ adapter: new Adapter() });

describe("History View", () => {
  it("should should display History View", () => {
    const now = new Date().toLocaleString("en-US", { timeZone: "UTC" });
    const mockHistory = [
      { message: "first commit", date: now },
      { message: "second commit", date: now },
      { message: "third commit", date: now }
    ];

    const wrapper = shallow(<History history={mockHistory} />);
    expect(wrapper.find("CheckpointCard")).toHaveLength(mockHistory.length);
    mockHistory.forEach(({ message }) => expect(wrapper.find(message)));
  });
});
