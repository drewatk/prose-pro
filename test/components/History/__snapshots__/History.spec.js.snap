// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History View should should display History View 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
>
  <Connect(CheckpointCard)
    key="0"
    timestamp="1/1/1970, 12:00:10 AM"
    version="first commit"
  />
  <Connect(CheckpointCard)
    key="1"
    timestamp="1/1/1970, 12:00:10 AM"
    version="second commit"
  />
  <Connect(CheckpointCard)
    key="2"
    timestamp="1/1/1970, 12:00:10 AM"
    version="third commit"
  />
</div>
`;
